module/builder/collection/builder.go

```diff
@@ -195,13 +195,29 @@ func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) er
 		minRefID := refChainFinalizedID
 
 		var transactions []*flow.TransactionBody
+		var totalByteSize uint
+		var totalGas uint64
 		for _, tx := range b.transactions.All() {
 
 			// if we have reached maximum number of transactions, stop
 			if uint(len(transactions)) >= b.config.MaxCollectionSize {
 				break
 			}
 
+			// we can break cause
+			// max byte size per tx << the max collection byte size
+			// to make it more effective in the future we can continue adding smaller ones
```

Comment:
I'm concerned by this comment:

In the worst case, if the byte size of the first transaction returned by `transactions.All()` is bigger than the collection's max byte size, then no transaction will be added. 

Could we add a check to say if the single transaction's byte size is bigger than the max colleciton byte size, we will `continue` instead of `break`:

```

			if tx.ByteSize() > b.config.MaxCollectionByteSize {
				continue
			}
			if totalByteSize+tx.ByteSize() > b.config.MaxCollectionByteSize {
				break
			}
```

This could act as a "filter".

Same comment to the `GasLimit`

Reply:
I think we covered this case by setting the tx max limit to 64K, but you're right there is no harm covering this edge case in case one day to remove tx size limits

Reply:
Finally we get to coin-selection problem!

Reply:
applied the suggested changes.